src/v/kafka/server/quota_manager.h

```diff
@@ -54,11 +56,26 @@ class quota_manager : public ss::peering_sharded_service<quota_manager> {
 public:
     using clock = ss::lowres_clock;
 
-    struct throttle_delay {
-        clock::duration duration{0};
+    // Accounting for quota on per-client and per-client-group basis
+    // last_seen_ms: used for gc keepalive
+    // tp_produce_rate: produce throughput tracking
+    // tp_fetch_rate: fetch throughput tracking
+    // pm_rate: partition mutation quota tracking
+    struct client_quota {
+        clock::time_point last_seen;
+        std::optional<atomic_token_bucket> tp_produce_rate;
+        std::optional<atomic_token_bucket> tp_fetch_rate;
+        std::optional<atomic_token_bucket> pm_rate;
     };
 
-    quota_manager();
+    using client_quotas_map_t = absl::node_hash_map<
+      ss::sstring,
+      ss::lw_shared_ptr<client_quota>,
+      sstring_hash,
+      sstring_eq>;
+    using client_quotas_t = ssx::sharded_ptr<client_quotas_map_t>;
+
+    explicit quota_manager(client_quotas_t& client_quotas);
```

Comment:
Just want to restate the argument for why the mutable `lw_shared` is safe here: off-owner mutation is always done through two layers of indirection - an iterator into the underlying map, which holds a _reference_ to the `lw_shared_ptr`, which is then deref'ed into a mutable `client_quota&` and passed to the mutation callback. Do I have that right?

Q: when will the ref count on the `lw_shared_ptr<client_quota>` (the quota map value itself) be `>1`? Only when you hold a copy of the map itself in an update func for `sharded_ptr`?

Reply:
Yes, that's all correct.

Reply:
nice 👍